src/lib/Parser.hs

```diff
@@ -260,6 +264,52 @@ topLet = do
         k = if v == mkName "eff" then EffectRowKind else TypeKind
         uTyKind = WithSrc (Just pos) $ UPrimExpr $ TCExpr k
 
+interfaceDef :: Parser [UDecl]
+interfaceDef = do
+  keyWord InterfaceKW
+  -- TODO: allow multi-parameter type classes
+  interfaceName <- upperName <|> symName
+  (oneAnnBinder, pos) <- withPos annBinder
+  keyWord WhereKW
+  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
+  let consName = mkInterfaceConsName interfaceName
+  let (varTypeName, oneAnnBinder') = case oneAnnBinder of
+        Bind v    -> (varName v, oneAnnBinder)
+        Ignore ty -> let genVar = mkName $ mkNoShadowingStr "gen" in
+            (genVar, Bind (genVar:>ty))
+  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder']
+  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
+  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  where
+    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
+    mkFunDefs meta ~(WithSrc _ (URecordTy (NoExt (LabeledItems items)))) =
+        fmap (\(name, ty :| []) -> mkOneFunDef meta (name, ty)) $ M.toList items
+    mkOneFunDef :: (SrcPos, Name, Name) -> (Label, UExpr) -> UDecl
+    mkOneFunDef (pos, varTypeName, interfaceName) (fLabel, fType) =
+      let (p, ann) = lhs
+      in let (ann', rhs') = addImplicitImplicitArgs pos ann rhs
+      in ULet PlainLet (p, ann') rhs'
+      where
+        lhs = let uAnnBinder = Bind $ instanceName :> ns (UApp
+                    (PlainArrow ()) (var interfaceName) (var varTypeName))
+          in (patb fLabel, Just $ ns $ UPi uAnnBinder ClassArrow fType)
+        rhs =
+          let recordStr = mkNoShadowingStr "recordVar"
+              recordPat = ns $ UPatRecord $ Ext (labeledSingleton fLabel (patb
+                fLabel)) $ Just (ns (UPatBinder (Ignore ())))
+              conPat = ns $ UPatCon (mkInterfaceConsName interfaceName)
+                $ toNest [patb recordStr]
+              let1 = ULet PlainLet (conPat, Nothing) $ var instanceName
+              let2 = ULet PlainLet (recordPat, Nothing) $ var $ mkName recordStr
+              body = ns $ UDecl let1 (ns $ UDecl let2 (var (mkName fLabel)))
+          in ns $ ULam (patb instanceStr, Nothing) ClassArrow body
+        
+        ns = WithSrc Nothing
+        patb s = ns $ UPatBinder $ Bind $ mkName s :> ()
+        instanceStr = mkNoShadowingStr "instance"
+        instanceName = mkName instanceStr
+        var name = ns $ UVar $ name :> ()
```

Comment:
I think this would be clearer if you combined the two lets and the where into a single `let` or `where` block. Is there a reason for all of the nesting?

Reply:
The second where is useful to avoid having to explicitly pass the parameters of  `mkOneFunDef` to `lhs` and `rhs, but maybe that is not such a big deal and reducing nesting should be preferred. WDYT? Re. the lets, that is an overlook, thanks:)